src/coreclr/jit/valuenum.cpp

```diff
@@ -7273,6 +7269,29 @@ void Compiler::fgValueNumberBlockAssignment(GenTree* tree)
                     isNewUniq = true;
                 }
 
+                if (!isNewUniq && (rhsVarDsc != nullptr) && (lhsVarDsc != nullptr))
```

Comment:
the check `isNewUniq` is weak here because it is possible that we have already generated a unique VN for rhs but did not report it here.
Because of that we can't narrow the cases that go inside this block to `assert(ClassLayout::AreCompatible(rhsVarDsc->GetLayout(), lhsVarDsc->GetLayout()))`.
However, all cases that pass all these "if" should have unique VN, we just ask for it twice for some trees.